### PR TITLE
Featured workspaces tab [GAWB-2524]

### PR DESCRIPTION
**Pull request by aednichols**
Requires https://github.com/broadinstitute/firecloud-develop/pull/657, https://github.com/broadinstitute/firecloud-develop/pull/653, https://github.com/broadinstitute/firecloud-develop/pull/651

To test, you will need to edit the `featured.json` configuration file in the appropriate bucket for the environment. Attached to the JIRA ticket, is a Cyberduck config shortcut to do so for dev.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/cljs/main/broadfcui/page/workspaces_list.cljs [hunk 1]
```diff
@@ -226,7 +226,12 @@
                                 (or (not public)
                                     (common/access-greater-than-equal-to? accessLevel "WRITER")))}
                   {:label "Public Workspaces"
-                   :predicate :public}]}
+                   :predicate :public}
+                  {:label "Featured Workspaces"
+                   :predicate (fn [{:keys [workspace]}]
+                                (= (count (filterv (fn [featured-workspace]
+                                  (and (= (:name workspace) (:name featured-workspace))
+                                        (= (:namespace workspace) (:namespace featured-workspace)))) (:featured-workspaces props))) 1))}]}
```
Comment (MattPutnam) [inline src/cljs/main/broadfcui/page/workspaces_list.cljs]: Convert `featured-workspaces` to a set upon loading, then this predicate can just be:
```clojure
(fn [workspace]
  (contains? (:featured-workspaces props) (ws-common/workspace->id workspace)))
```

File: src/cljs/main/broadfcui/page/workspaces_list.cljs [hunk 2]
```diff
@@ -402,15 +407,16 @@
...
        (net/render-with-ajax
         (:workspaces-response server-response)
         (fn []
           [WorkspaceTable
            (assoc props
              :workspaces workspaces
              :billing-projects billing-projects
-             :disabled-reason disabled-reason)])
+             :disabled-reason disabled-reason
+             :featured-workspaces featured-workspaces)])
```
Comment (MattPutnam) [inline src/cljs/main/broadfcui/page/workspaces_list.cljs]: This `assoc` is getting silly, let's convert to:
```clojure
(merge props (utils/restructure workspaces billing-projects disabled-reason featured-workspaces))
```

File: src/cljs/main/broadfcui/page/workspaces_list.cljs [hunk 3]
```diff
@@ -426,7 +432,10 @@
                  :error-message err-text :disabled-reason :error)
           (swap! state update :server-response assoc
                  :billing-projects (map :projectName projects)
-                 :disabled-reason (if (empty? projects) :no-billing nil))))))})
+                 :disabled-reason (if (empty? projects) :no-billing nil)))))
+     (utils/ajax
+      {:url (config/featured-json-url)
+       :on-done (fn [{:keys [get-parsed-response]}] (swap! state update :server-response assoc :featured-workspaces (get-parsed-response)))}))})
```
Comment (MattPutnam) [inline src/cljs/main/broadfcui/page/workspaces_list.cljs]: Newline after the function arguments

File: src/cljs/main/broadfcui/page/workspaces_list.cljs (hunk 1 above)
Comment (aednichols) [reply src/cljs/main/broadfcui/page/workspaces_list.cljs]: Thanks, this is really nice.

I have to rein in the part of my brain that wants to write this:
```
@Override
public boolean equals(WorkspaceRef other) {
    return this.name.equals(other.name) && this.namespace.equals(other.namespace);
}
```

**Comment by davidangb** [comment]
looks like you made firecloud-develop PRs for local, FiaB, and alpha ... we actually need local, FiaB, and *dev*, not alpha.

File: src/cljs/main/broadfcui/page/workspaces_list.cljs [hunk 4]
```diff
@@ -426,7 +426,11 @@
                  :error-message err-text :disabled-reason :error)
           (swap! state update :server-response assoc
                  :billing-projects (map :projectName projects)
-                 :disabled-reason (if (empty? projects) :no-billing nil))))))})
+                 :disabled-reason (if (empty? projects) :no-billing nil)))))
+     (utils/ajax
+      {:url (config/featured-json-url)
+       :on-done (fn [{:keys [get-parsed-response]}]
+                  (swap! state update :server-response assoc :featured-workspaces (set (get-parsed-response))))}))})
```
Comment (davidangb) [inline src/cljs/main/broadfcui/page/workspaces_list.cljs]: I think this ajax call should be pretty resilient in the case that the user editing/uploading the featured.json made a mistake. Could you trap for non-success responses and for malformed json, and in those cases use the empty set? If this is a bunch of work, I'd be okay calling it scope creep and not doing it.

File: src/cljs/main/broadfcui/page/workspaces_list.cljs [hunk 5]
```diff
@@ -226,7 +226,10 @@
                                 (or (not public)
                                     (common/access-greater-than-equal-to? accessLevel "WRITER")))}
                   {:label "Public Workspaces"
-                   :predicate :public}]}
+                   :predicate :public}
+                  {:label "Featured Workspaces"
+                   :predicate (fn [workspace]
+                                (contains? (:featured-workspaces props) (ws-common/workspace->id workspace)))}]}
```
Comment (davidangb) [inline src/cljs/main/broadfcui/page/workspaces_list.cljs]: This predicate is logically correct, but it has slowed down browser performance on the workspace list when switching tabs - to an extent that I can see it with the naked eye. Look at the `attach-table-data` function earlier in this file. I think we may need to pre-calculate a `featured?` key for the workspace during `attach-table-data`, and use that key in the predicate.

File: src/cljs/main/broadfcui/common/notifications.cljs [hunk 6]
```diff
@@ -71,7 +71,6 @@
    :-load-service-alerts
    (fn [{:keys [this]} & [first-time?]]
      (utils/ajax {:url (config/alerts-json-url)
-                  :headers {"Cache-Control" "no-store, no-cache"}
```
Comment (aednichols) [inline src/cljs/main/broadfcui/common/notifications.cljs:74]: Removed because this has no effect and was misleading during debugging
> Note: Cache-Control is also a header you can specify in HTTP requests for an object. Cloud Storage ignores such headers if they are set to avoid cached content.

[Source.](https://cloud.google.com/storage/docs/metadata?hl=en_US&_ga=2.170911328.-1166978692.1449000140#cache-control)

File: src/cljs/main/broadfcui/page/workspaces_list.cljs [hunk 7]
```diff
@@ -19,7 +19,7 @@
    [broadfcui.page.workspace.create :as create]
    [broadfcui.persistence :as persistence]
    [broadfcui.utils :as utils]
-   ))
+   [broadfcui.page.workspace.workspace-common :as ws-common]))
```
Comment (MattPutnam) [inline src/cljs/main/broadfcui/page/workspaces_list.cljs]: Unused import

File: src/cljs/main/broadfcui/page/workspaces_list.cljs [hunk 8]
```diff
@@ -189,7 +189,8 @@
                     (if (contains? domain-groups config/tcga-authorization-domain)
                       tcga-disabled-text
                       non-dbGap-disabled-text))
-      :restricted? (seq auth-domain))))
+      :restricted? (seq auth-domain)
+      :featured?  (contains? featured-workspaces (select-keys workspace [:namespace :name])))))
```
Comment (MattPutnam) [inline src/cljs/main/broadfcui/page/workspaces_list.cljs]: You already have `(select-keys workspace [:namespace :name])` defined as `workspace-id`

**Review by MattPutnam — APPROVED** [review]
👍 with two small bits of cleanup

**Review by davidangb — APPROVED** [review]
looks good to me, thanks! :+1:
